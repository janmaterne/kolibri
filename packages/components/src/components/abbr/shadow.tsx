import type { JSX } from '@stencil/core';
import { Component, h, Host, Prop, State, Watch } from '@stencil/core';
import type { AbbrAPI, AbbrStates, LabelPropType } from '../../schema';
import { validateLabel } from '../../schema';

/**
 * @slot - The abbreviation (short form).
 */
@Component({
	tag: 'kol-abbr',
	styleUrls: {
		default: './style.scss',
	},
	shadow: true,
})
export class KolAbbr implements AbbrAPI {
	public render(): JSX.Element {
		return (
			<Host class="kol-abbr">
				<abbr>
					<slot />
				</abbr>
				{this.state._label ? ` (${this.state._label})` : ''}
			</Host>
		);
	}

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.).
	 */
	@Prop() public _label?: LabelPropType;

	@State() public state: AbbrStates = {
		_label: '', // âš  required
	};

	@Watch('_label')
	public validateLabel(value?: LabelPropType): void {
		validateLabel(this, value, {
			required: true,
		});
	}

	public componentWillLoad(): void {
		this.validateLabel(this._label);
	}
}
