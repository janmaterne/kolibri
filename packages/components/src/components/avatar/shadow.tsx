import type { JSX } from '@stencil/core';
import { Component, h, Prop } from '@stencil/core';

import type { AvatarProps } from '../../schema';
import { KolAvatarWcTag } from '../../core/component-names';

@Component({
	tag: 'kol-avatar',
	styleUrls: {
		default: './style.scss',
	},
	shadow: true,
})
export class KolAvatar implements AvatarProps {
	public render(): JSX.Element {
		return <KolAvatarWcTag _src={this._src} _label={this._label}></KolAvatarWcTag>;
	}

	/**
	 * Sets the image `src` attribute to the given string.
	 */
	@Prop() public _src?: string;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.).
	 */
	@Prop() public _label!: string;
}
