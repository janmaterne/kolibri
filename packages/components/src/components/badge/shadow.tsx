import type { BadgeAPI, BadgeStates, ButtonProps, KoliBriIconsProp, LabelPropType, PropColor, Stringified } from '../../schema';
import { featureHint, handleColorChange, objectObjectHandler, parseJson, setState, validateColor, validateIcons } from '../../schema';
import { Component, h, Host, Prop, State, Watch } from '@stencil/core';
import { KolSpanFc } from '@fc';

import { nonce } from '../../utils/dev.utils';

import type { JSX } from '@stencil/core';
import { KolButtonWcTag } from '../../core/component-names';
featureHint(`[KolBadge] Optimierung des _color-Properties (rgba, rgb, hex usw.).`);

@Component({
	tag: 'kol-badge',
	styleUrls: {
		default: './style.scss',
	},
	shadow: true,
})
export class KolBadge implements BadgeAPI {
	private bgColorStr = '#000';
	private colorStr = '#fff';
	private readonly id = nonce();

	private renderSmartButton(props: ButtonProps): JSX.Element {
		return (
			<KolButtonWcTag
				_ariaControls={this.id}
				_customClass={props._customClass}
				_disabled={props._disabled}
				_hideLabel={true}
				_icons={props._icons}
				_id={props._id}
				_label={props._label}
				_on={props._on}
				_tooltipAlign={props._tooltipAlign}
				_variant={props._variant}
			></KolButtonWcTag>
		);
	}

	public render(): JSX.Element {
		const hasSmartButton = typeof this.state._smartButton === 'object' && this.state._smartButton !== null;

		return (
			<Host class="kol-badge">
				<span
					class={{
						'smart-button': typeof this.state._smartButton === 'object' && this.state._smartButton !== null,
					}}
					style={{
						backgroundColor: this.bgColorStr,
						color: this.colorStr,
					}}
				>
					<KolSpanFc id={hasSmartButton ? this.id : undefined} allowMarkdown icons={this.state._icons} label={this._label} />
					{hasSmartButton && this.renderSmartButton(this.state._smartButton as ButtonProps)}
				</span>
			</Host>
		);
	}

	/**
	 * Defines the backgroundColor and foregroundColor.
	 */
	@Prop() public _color?: Stringified<PropColor> = '#000';

	/**
	 * Defines the icon classnames (e.g. `_icons="fa-solid fa-user"`).
	 */
	@Prop() public _icons?: Stringified<KoliBriIconsProp>;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.).
	 */
	@Prop() public _label!: LabelPropType;

	/**
	 * Allows to add a button with an arbitrary action within the element (_hide-label only).
	 */
	@Prop() public _smartButton?: Stringified<ButtonProps>;

	@State() public state: BadgeStates = {
		_color: {
			backgroundColor: '#000',
			foregroundColor: '#fff',
		},
		_icons: {},
	};

	private handleColorChange = (value: unknown) => {
		const colorPair = handleColorChange(value);
		this.bgColorStr = colorPair.backgroundColor;
		this.colorStr = colorPair.foregroundColor as string;
	};

	@Watch('_icons')
	public validateIcons(value?: KoliBriIconsProp): void {
		validateIcons(this, value);
	}

	@Watch('_color')
	public validateColor(value?: Stringified<PropColor>): void {
		validateColor(this, value, {
			defaultValue: '#000',
			hooks: {
				beforePatch: this.handleColorChange,
			},
		});
	}

	@Watch('_smartButton')
	public validateSmartButton(value?: ButtonProps | string): void {
		objectObjectHandler(value, () => {
			try {
				value = parseJson<ButtonProps>(value as string);
				// eslint-disable-next-line no-empty
			} catch (e) {
				// value behält den ursprünglichen Wert
			}
			setState(this, '_smartButton', value);
		});
	}

	public componentWillLoad(): void {
		this.validateIcons(this._icons);
		this.validateColor(this._color);
		this.validateSmartButton(this._smartButton);
	}
}
