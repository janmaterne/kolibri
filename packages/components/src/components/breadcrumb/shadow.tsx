import type { BreadcrumbAPI, BreadcrumbLinkProps, BreadcrumbStates, LabelPropType, LinkProps, Stringified } from '../../schema';
import { a11yHintLabelingLandmarks, validateLabel } from '../../schema';
import { Component, Fragment, h, Prop, State, Watch } from '@stencil/core';

import { addNavLabel, removeNavLabel } from '../../utils/unique-nav-labels';
import { watchNavLinks } from '../nav/validation';

import type { JSX } from '@stencil/core';
import { KolIconTag, KolLinkTag } from '../../core/component-names';

@Component({
	tag: 'kol-breadcrumb',
	styleUrls: {
		default: './style.scss',
	},
	shadow: true,
})
export class KolBreadcrumb implements BreadcrumbAPI {
	private readonly renderLink = (link: BreadcrumbLinkProps, index: number): JSX.Element => {
		const lastIndex = this.state._links.length - 1;
		return (
			<li class="kol-breadcrumb__list-element" key={index}>
				{index !== 0 && <KolIconTag class="kol-breadcrumb__icon" _label="" _icons="codicon codicon-chevron-right" />}
				{index === lastIndex ? (
					<span class="kol-breadcrumb__list-element-span">
						{link._hideLabel ? (
							<KolIconTag
								class="kol-breadcrumb__icon"
								_label={link._label}
								_icons={typeof link._icons === 'string' ? link._icons : 'codicon codicon-symbol-event'}
							/>
						) : (
							<>{link._label}</>
						)}
					</span>
				) : (
					<KolLinkTag class="kol-breadcrumb__link" {...link}></KolLinkTag>
				)}
			</li>
		);
	};

	public render(): JSX.Element {
		return (
			<nav class="kol-breadcrumb" aria-label={this.state._label}>
				<ul class="kol-breadcrumb__list">
					{this.state._links.length === 0 && (
						<li>
							<KolIconTag class="kol-breadcrumb_icon" _label="" _icons="codicon codicon-home" />…
						</li>
					)}
					{this.state._links.map(this.renderLink)}
				</ul>
			</nav>
		);
	}

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.).
	 */
	@Prop() public _label!: LabelPropType;

	/**
	 * Defines the list of links combined with their labels to render.
	 */
	@Prop() public _links!: Stringified<BreadcrumbLinkProps[]>;

	@State() public state: BreadcrumbStates = {
		_label: '', // ⚠ required
		_links: [],
	};

	@Watch('_label')
	public validateLabel(value?: LabelPropType, _oldValue?: LabelPropType, initial = false): void {
		if (!initial) {
			removeNavLabel(this.state._label); // remove the current
		}
		validateLabel(this, value, {
			required: true,
		});
		a11yHintLabelingLandmarks(value);
		addNavLabel(this.state._label); // add the state instead of prop, because the prop could be invalid and not set as new label
	}

	@Watch('_links')
	public validateLinks(value?: Stringified<LinkProps[]>): void {
		watchNavLinks('KolBreadcrumb', this, value);
	}

	public componentWillLoad(): void {
		this.validateLabel(this._label, undefined, true);
		this.validateLinks(this._links);
	}

	public disconnectedCallback(): void {
		removeNavLabel(this.state._label);
	}
}
