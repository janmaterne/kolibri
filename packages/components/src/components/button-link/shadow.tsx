import { Component, Element, h, Host, JSX, Prop } from '@stencil/core';

import { Stringified } from '../../types/common';
import { AlternativeButtonLinkRolePropType } from '../../types/props/alternative-button-link-role';
import { AriaCurrentPropType } from '../../types/props/aria-current';
import { ButtonCallbacksPropType } from '../../types/props/button-callbacks';
import { ButtonTypePropType } from '../../types/props/button-type';
import { IconsPropType } from '../../types/props/icons';
import { IdPropType } from '../../types/props/id';
import { LabelWithExpertSlotPropType } from '../../types/props/label';
import { NamePropType } from '../../types/props/name';
import { SyncValueBySelectorPropType } from '../../types/props/sync-value-by-selector';
import { TooltipAlignPropType } from '../../types/props/tooltip-align';
import { StencilUnknown } from '../../types/unknown';
import { propagateFocus } from '../../utils/reuse';
import { Props } from './types';
import { KolButtonWcTag } from '../../core/component-names';

@Component({
	tag: 'kol-button-link',
	styleUrls: {
		default: './style.scss',
	},
	shadow: true,
})
export class KolButtonLink implements Props {
	@Element() private readonly host?: HTMLKolButtonLinkElement;

	private readonly catchRef = (ref?: HTMLKolButtonWcElement) => {
		propagateFocus(this.host, ref);
	};

	public render(): JSX.Element {
		return (
			<Host class="kol-button-link">
				<KolButtonWcTag
					ref={this.catchRef}
					_accessKey={this._accessKey}
					_ariaControls={this._ariaControls}
					_ariaCurrent={this._ariaCurrent}
					_ariaExpanded={this._ariaExpanded}
					_ariaLabel={this._ariaLabel}
					_ariaSelected={this._ariaSelected}
					_disabled={this._disabled}
					_icons={this._icons || this._icon}
					_hideLabel={this._hideLabel || this._iconOnly}
					_id={this._id}
					_label={this._label}
					_name={this._name}
					_on={this._on}
					_role="link"
					_syncValueBySelector={this._syncValueBySelector}
					_tabIndex={this._tabIndex}
					_tooltipAlign={this._tooltipAlign}
					_type={this._type}
					_value={this._value}
				>
					<slot name="expert" slot="expert"></slot>
				</KolButtonWcTag>
			</Host>
		);
	}

	/**
	 * Defines which key combination can be used to trigger or focus the interactive element of the component.
	 */
	@Prop() public _accessKey?: string;

	/**
	 * Defines which elements are controlled by this component. (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-controls)
	 */
	@Prop() public _ariaControls?: string;

	/**
	 * Deprecated: Marks the element as the selected in a group of related elements. Can be one of the following: `date` | `location` | `page` | `step` | `time` | `true`. (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-current)
	 *
	 * @deprecated aria-current is not necessary for buttons. will be removed in version 2.
	 */
	@Prop() public _ariaCurrent?: AriaCurrentPropType;

	/**
	 * Defines whether the interactive element of the component expanded something. (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded)
	 * @TODO: Change type back to `AriaExpandedPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _ariaExpanded?: boolean;

	/**
	 * Deprecated: Setzt die semantische Beschriftung der Komponente.
	 *
	 * @deprecated use _label instead
	 */
	@Prop({ mutable: true, reflect: false }) public _ariaLabel?: string;

	/**
	 * Defines whether the interactive element of the component is selected (e.g. role=tab). (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-selected)
	 * @TODO: Change type back to `AriaSelectedPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _ariaSelected?: boolean;

	/**
	 * Makes the element not focusable and ignore all events.
	 * @TODO: Change type back to `DisabledPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _disabled?: boolean = false;

	/**
	 * Hides the caption by default and displays the caption text with a tooltip when the
	 * interactive element is focused or the mouse is over it.
	 * @TODO: Change type back to `HideLabelPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _hideLabel?: boolean = false;

	/**
	 * @deprecated Use _icons.
	 */
	@Prop() public _icon?: IconsPropType;

	/**
	 * Defines the icon classnames (e.g. `_icons="fa-solid fa-user"`).
	 */
	@Prop() public _icons?: IconsPropType;

	/**
	 * Deprecated: Hides the label and shows the description in a Tooltip instead.
	 * @deprecated use _hide-label
	 */
	@Prop() public _iconOnly?: boolean;

	/**
	 * Defines the internal ID of the primary component element.
	 */
	@Prop() public _id?: IdPropType;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.). Set to `false` to enable the expert slot.
	 */
	@Prop() public _label!: LabelWithExpertSlotPropType;

	/**
	 * Defines the technical name of an input field.
	 */
	@Prop() public _name?: NamePropType;

	/**
	 * Gibt die EventCallback-Funktionen f√ºr die Button-Events an.
	 */
	@Prop() public _on?: ButtonCallbacksPropType<StencilUnknown>;

	/**
	 * Defines the role of the components primary element.
	 */
	@Prop() public _role?: AlternativeButtonLinkRolePropType;

	/**
	 * Selector for synchronizing the value with another input element.
	 * @internal
	 */
	@Prop() public _syncValueBySelector?: SyncValueBySelectorPropType;

	/**
	 * Defines which tab-index the primary element of the component has. (https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex)
	 */
	@Prop() public _tabIndex?: number;

	/**
	 * Defines where to show the Tooltip preferably: top, right, bottom or left.
	 */
	@Prop() public _tooltipAlign?: TooltipAlignPropType = 'top';

	/**
	 * Defines either the type of the component or of the components interactive element.
	 */
	@Prop() public _type?: ButtonTypePropType = 'button';

	/**
	 * Defines the value that the button emits on click.
	 */
	@Prop() public _value?: Stringified<StencilUnknown>;
}
