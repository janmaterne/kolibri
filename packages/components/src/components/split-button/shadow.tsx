import { Component, h, Host, JSX, Prop, State, Watch } from '@stencil/core';

import { Stringified } from '../../types/common';
import { AlternativeButtonLinkRolePropType } from '../../types/props/alternative-button-link-role';
import { ButtonCallbacksPropType } from '../../types/props/button-callbacks';
import { ButtonTypePropType } from '../../types/props/button-type';
import { ButtonVariantPropType } from '../../types/props/button-variant';
import { CustomClassPropType } from '../../types/props/custom-class';
import { IconsPropType } from '../../types/props/icons';
import { LabelPropType } from '../../types/props/label';
import { validateShow } from '../../types/props/show';
import { SyncValueBySelectorPropType } from '../../types/props/sync-value-by-selector';
import { TooltipAlignPropType } from '../../types/props/tooltip-align';
import { StencilUnknown } from '../../types/unknown';
import { API, States } from './types';
import { KolButtonWcTag } from '../../core/component-names';

/**
 * @slot - Ermöglicht das Einfügen beliebigen HTMLs in das dropdown.
 */
@Component({
	tag: 'kol-split-button',
	styleUrls: {
		default: './style.scss',
	},
	shadow: true,
})
export class KolSplitButton implements API {
	private dropdown: HTMLDivElement | undefined;
	private dropdownContent: HTMLDivElement | undefined;

	private readonly clickButtonHandler = {
		onClick: (e: MouseEvent) => {
			if (typeof this._on?.onClick === 'function') {
				// TODO: this._on is not validated
				this._on?.onClick(e, this._value);
			} else {
				this.toggleDropdown();
			}
		},
	};
	private readonly clickToggleHandler = { onClick: () => this.toggleDropdown() };

	private readonly openDropdown = () => {
		if (this.dropdown && this.dropdownContent) {
			this.dropdown.style.height = `${this.dropdownContent.clientHeight}px`;
			this.state = { ...this.state, _show: true };
		}
	};
	private readonly closeDropdown = () => {
		if (this.dropdown && this.dropdownContent) {
			this.dropdown.style.height = ``;
			this.state = { ...this.state, _show: false };
		}
	};
	private readonly toggleDropdown = (value?: boolean) => {
		const openIt = typeof value === 'boolean' ? value : !this.state._show;
		if (openIt) this.openDropdown();
		else this.closeDropdown();
	};
	private forceCounter = 100; // because the dropdown could be empty
	private readonly forceOpenDropdown = () => {
		if (this.forceCounter > 0) {
			if (this.dropdownContent?.clientHeight) this.openDropdown();
			else setTimeout(this.forceOpenDropdown, 10);
			this.forceCounter--;
		}
	};

	private readonly catchDropdownElements = (e?: HTMLDivElement | null) => {
		if (e) {
			this.dropdown = e;
			setTimeout(() => {
				this.dropdownContent = e.firstChild as HTMLDivElement;
				if (this._show) this.forceOpenDropdown();
			});
		}
	};

	public render(): JSX.Element {
		return (
			<Host class="kol-split-button split-button">
				<KolButtonWcTag
					class={{
						'main-button': true,
						button: true,
						[this._variant as string]: this._variant !== 'custom',
						[this._customClass as string]: this._variant === 'custom' && typeof this._customClass === 'string' && this._customClass.length > 0,
					}}
					_accessKey={this._accessKey}
					_ariaControls={this._ariaControls}
					_ariaExpanded={this._ariaExpanded}
					_ariaSelected={this._ariaSelected}
					_customClass={this._customClass}
					_disabled={this._disabled}
					_icons={this._icons || this._icon}
					_hideLabel={this._hideLabel}
					_label={this._label}
					_name={this._name}
					_on={this.clickButtonHandler}
					_role={this._role}
					_syncValueBySelector={this._syncValueBySelector}
					_tabIndex={this._tabIndex}
					_tooltipAlign={this._tooltipAlign}
					_type={this._type}
					_value={this._value}
					_variant={this._variant}
				></KolButtonWcTag>
				<div class="horizontal-line"></div>
				<KolButtonWcTag
					class="secondary-button"
					_disabled={this._disabled}
					_hideLabel
					_icons="codicon codicon-triangle-down"
					_label={`dropdown ${this.state._show ? 'schließen' : 'öffnen'}`} // @todo: translate
					_on={this.clickToggleHandler}
				></KolButtonWcTag>
				<div class="popover" ref={this.catchDropdownElements}>
					<div class="popover-content">
						<slot />
					</div>
				</div>
			</Host>
		);
	}

	/**
	 * Defines which key combination can be used to trigger or focus the interactive element of the component.
	 */
	@Prop() public _accessKey?: string;

	/**
	 * Defines which elements are controlled by this component. (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-controls)
	 */
	@Prop() public _ariaControls?: string;

	/**
	 * Defines whether the interactive element of the component expanded something. (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded)
	 */
	@Prop() public _ariaExpanded?: boolean;

	/**
	 * Deprecated: Setzt die semantische Beschriftung der Komponente.
	 *
	 * @deprecated use _label instead
	 */
	@Prop() public _ariaLabel?: string;

	/**
	 * Defines whether the interactive element of the component is selected (e.g. role=tab). (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-selected)
	 */
	@Prop() public _ariaSelected?: boolean;

	/**
	 * Defines the custom class attribute if _variant="custom" is set.
	 */
	@Prop() public _customClass?: CustomClassPropType;

	/**
	 * Makes the element not focusable and ignore all events.
	 */
	@Prop() public _disabled?: boolean = false;

	/**
	 * Hides the caption by default and displays the caption text with a tooltip when the
	 * interactive element is focused or the mouse is over it.
	 * @TODO: Change type back to `HideLabelPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _hideLabel?: boolean = false;

	/**
	 * @deprecated Use _icons.
	 */
	@Prop() public _icon?: IconsPropType;

	/**
	 * Defines the icon classnames (e.g. `_icons="fa-solid fa-user"`).
	 */
	@Prop() public _icons?: IconsPropType;

	/**
	 * Defines the internal ID of the primary component element.
	 */
	@Prop() public _id?: string;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.).
	 */
	@Prop() public _label!: LabelPropType;

	/**
	 * Defines the technical name of an input field.
	 */
	@Prop() public _name?: string;

	/**
	 * Defines the callback functions for button events.
	 */
	@Prop() public _on?: ButtonCallbacksPropType<StencilUnknown>;

	/**
	 * Defines the role of the components primary element.
	 */
	@Prop() public _role?: AlternativeButtonLinkRolePropType;

	/**
	 * Selector for synchronizing the value with another input element.
	 * @internal
	 */
	@Prop() public _syncValueBySelector?: SyncValueBySelectorPropType;

	/**
	 * Makes the element show up.
	 * @TODO: Change type back to `ShowPropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _show?: boolean = false;

	/**
	 * Deprecated: Defines whether to show the dropdown menu.
	 * @deprecated use _show instead
	 */
	@Prop({ mutable: true, reflect: true }) public _showDropdown?: boolean = false;

	/**
	 * Defines which tab-index the primary element of the component has. (https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex)
	 */
	@Prop() public _tabIndex?: number;

	/**
	 * Defines where to show the Tooltip preferably: top, right, bottom or left.
	 */
	@Prop() public _tooltipAlign?: TooltipAlignPropType = 'top';

	/**
	 * Defines either the type of the component or of the components interactive element.
	 */
	@Prop() public _type?: ButtonTypePropType = 'button';

	/**
	 * Defines the value that the button emits on click.
	 */
	@Prop() public _value?: Stringified<StencilUnknown>;

	/**
	 * Defines which variant should be used for presentation.
	 */
	@Prop() public _variant?: ButtonVariantPropType = 'normal';

	@State() public state: States = {
		_show: false,
	};

	/**
	 * @deprecated will be removed in the next major version
	 */
	@Watch('_show')
	public validateShowDropdown(value?: boolean): void {
		this.validateShow(value);
	}

	@Watch('_show')
	public validateShow(value?: boolean): void {
		validateShow(this, value, {
			hooks: {
				afterPatch: (value) => {
					this.toggleDropdown(!!value);
				},
			},
		});
	}

	public componentWillLoad(): void {
		this.validateShow(this._show || this._show);
	}
}
