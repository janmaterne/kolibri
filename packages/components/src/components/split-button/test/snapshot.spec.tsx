import { KolSplitButtonTag } from '../../../core/component-names';
import type { SplitButtonProps } from '../../../schema';
import { executeSnapshotTests } from '../../../utils/testing';

import { KolSplitButton } from '../shadow';

executeSnapshotTests<SplitButtonProps>(
	KolSplitButtonTag,
	[KolSplitButton],
	[
		{ _label: 'Label' },
		{ _label: 'Label', _hideLabel: true },
		{ _label: 'Label', _disabled: true },
		{ _label: 'Label', _name: 'Name' },
		{ _label: 'Label', _icons: 'codicon codicon-git-pull-request' },

		...['primary', 'secondary', 'normal', 'danger', 'ghost'].map(
			(_variant) =>
				({
					_label: 'Label',
					_variant,
				}) as SplitButtonProps,
		),
	],
);
