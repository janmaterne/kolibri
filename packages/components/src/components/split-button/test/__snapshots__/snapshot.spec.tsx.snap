// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kol-split-button should render with _label="Label" _disabled=true 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _disabled="" _label="Label" _tooltipalign="top" _type="button" _variant="normal" class="button main-button normal"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _disabled="" _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;

exports[`kol-split-button should render with _label="Label" _hideLabel=true 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _hidelabel="" _label="Label" _tooltipalign="top" _type="button" _variant="normal" class="button main-button normal"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;

exports[`kol-split-button should render with _label="Label" _icons="codicon codicon-git-pull-request" 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _icons="codicon codicon-git-pull-request" _label="Label" _tooltipalign="top" _type="button" _variant="normal" class="button main-button normal"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;

exports[`kol-split-button should render with _label="Label" _name="Name" 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _label="Label" _name="Name" _tooltipalign="top" _type="button" _variant="normal" class="button main-button normal"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;

exports[`kol-split-button should render with _label="Label" _variant="danger" 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _label="Label" _tooltipalign="top" _type="button" _variant="danger" class="button danger main-button"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;

exports[`kol-split-button should render with _label="Label" _variant="ghost" 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _label="Label" _tooltipalign="top" _type="button" _variant="ghost" class="button ghost main-button"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;

exports[`kol-split-button should render with _label="Label" _variant="normal" 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _label="Label" _tooltipalign="top" _type="button" _variant="normal" class="button main-button normal"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;

exports[`kol-split-button should render with _label="Label" _variant="primary" 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _label="Label" _tooltipalign="top" _type="button" _variant="primary" class="button main-button primary"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;

exports[`kol-split-button should render with _label="Label" _variant="secondary" 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _label="Label" _tooltipalign="top" _type="button" _variant="secondary" class="button main-button secondary"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;

exports[`kol-split-button should render with _label="Label" 1`] = `
<kol-split-button class="kol-split-button">
  <template shadowrootmode="open">
    <div class="split-button-root">
      <kol-button-wc _label="Label" _tooltipalign="top" _type="button" _variant="normal" class="button main-button normal"></kol-button-wc>
      <div class="horizontal-line"></div>
      <kol-button-wc _hidelabel="" _icons="codicon codicon-triangle-down" _label="kol-split-button-dropdown-label-open" class="secondary-button"></kol-button-wc>
    </div>
    <kol-popover-wc _align="bottom">
      <slot></slot>
    </kol-popover-wc>
  </template>
</kol-split-button>
`;
